pkg/imgpkg/bundle/locations.go

```diff
@@ -0,0 +1,175 @@
+// Copyright 2020 VMware, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+package bundle
+
+import (
+	"archive/tar"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"path/filepath"
+
+	"github.com/cppforlife/go-cli-ui/ui"
+	"github.com/google/go-containerregistry/pkg/name"
+	regv1 "github.com/google/go-containerregistry/pkg/v1"
+	"github.com/google/go-containerregistry/pkg/v1/remote/transport"
+	"github.com/google/go-containerregistry/pkg/v1/types"
+	"github.com/k14s/imgpkg/pkg/imgpkg/image"
+	"github.com/k14s/imgpkg/pkg/imgpkg/plainimage"
+	"github.com/k14s/imgpkg/pkg/imgpkg/util"
+)
+
+const (
+	locationsTag string = "%s-%s.locations.imgpkg"
+)
+
+type LocationsNotFound struct {
+	image string
+}
+
+func (n LocationsNotFound) Error() string {
+	return fmt.Sprintf("Locations image in %s could not be found", n.image)
+}
+
+type Locations struct {
+	reader LocationImageReader
+	logger util.LoggerWithLevels
+}
+
+type LocationImageReader interface {
+	Read(img regv1.Image) (LocationConfig, error)
+}
+
+func NewLocations(logger util.LoggerWithLevels) *Locations {
+	return NewLocationsWithReader(&locationsSingleLayerReader{}, logger)
+}
+
+func NewLocationsWithReader(reader LocationImageReader, logger util.LoggerWithLevels) *Locations {
+	return &Locations{reader: reader, logger: logger}
+}
+
+func (r Locations) Fetch(registry image.ImagesMetadata, bundleRef name.Digest) (LocationConfig, error) {
```

Comment:
should we move registry into a constructor?

Reply:
Unfortunately, Fetch and Save need 2 different types of registry struct. One needs the Reader the other needs the Writer. If we moved the registry to the constructor we would have to provide a reader/writer to accommodate both cases.
The problem with this is that the Bundle struct currently only has Read access to the registry when collecting the Image References aka `Bundle.AllImagesRefs`.

I decided to move the registry to the function itself because the only situation we need to write on this struct is in the NoteCopy function, so we would be giving write access to the majority of the cases where we do not need it.

Reply:
sounds good. makes sense.